app/scripts/lib/ComposableObservableStore.ts

```diff
@@ -0,0 +1,130 @@
+import { ObservableStore } from '@metamask/obs-store';
+import {
+  ActionConstraint,
+  ControllerMessenger,
+  EventConstraint,
+  getPersistentState,
+} from '@metamask/base-controller';
+import { getKnownPropertyNames } from '@metamask/utils';
+import {
+  BackgroundStateProxy,
+  MemStoreControllers,
+  MemStoreControllersComposedState,
+} from '../../../shared/types/metamask';
+
+/**
+ * An ObservableStore that can compose the state objects of its child stores and controllers
+ */
+export default class ComposableObservableStore extends ObservableStore<BackgroundStateProxy> {
+  /**
+   * Describes which stores are being composed. The key is the name of the
+   * store, and the value is either an ObservableStore, or a controller that
+   * extends one of the two base controllers in the `@metamask/base-controller`
+   * package.
+   */
+  config: Partial<MemStoreControllers> = {};
+  controllerMessenger: ControllerMessenger<ActionConstraint, EventConstraint>;
+  persist: boolean;
+
+  /**
+   * Create a new store
+   *
+   * @param options
+   * @param [options.config] - Map of internal state keys to child stores and controllers
+   * @param options.controllerMessenger - The controller messenger, used for subscribing to events from BaseControllerV2-based controllers.
+   * @param [options.state] - The composed state of the child stores and controllers
+   * @param [options.persist] - Whether or not to apply the persistence for v2 controllers
+   */
+  constructor({
+    config,
+    controllerMessenger,
+    state,
+    persist,
+  }: {
+    config: MemStoreControllers;
+    controllerMessenger: ControllerMessenger<ActionConstraint, EventConstraint>;
+    state: BackgroundStateProxy;
+    persist: boolean;
+  }) {
+    super(state);
+    this.persist = persist;
+    this.controllerMessenger = controllerMessenger;
+    if (config) {
+      this.updateStructure(config);
+    }
+  }
+
+  /**
+   * Composes a new internal store subscription structure
+   *
+   * @param {Record<string, object>} config - Describes which stores are being
+   *   composed. The key is the name of the store, and the value is either an
+   *   ObserableStore, or a controller that extends one of the two base
+   *   controllers in the `@metamask/base-controller` package.
+   */
+  updateStructure(config: MemStoreControllers) {
+    this.config = config;
+    this.removeAllListeners();
+    const initialState = {};
+    for (const controllerKey of getKnownPropertyNames(config)) {
+      const controller = config[controllerKey];
+      if (!controller) {
+        throw new Error(`Undefined '${controllerKey}'`);
+      }
+
+      if ('store' in controller && 'subscribe' in controller.store) {
```

Comment:
Minor, `Boolean(controller?.store?.subscribe)`?

Or is this to satisfy the typings to prevent them being detected as potentially undefined?

Reply:
This is necessary for the TypeScript conversion. 

`typeof controller` includes some controllers that have a `store` property and others that do not. Because of this, `controller` needs to be narrowed with a type guard before we can attempt to access the `store` property. Without this narrowing, we get the following error:

```ts
controller.store; // Property 'store' does not exist on type 'ControllersWithoutStore'.ts(2339)
const { store } = controller; // Same error
```

The `in` operator narrows the type of `controller` down to an object with a `store` property.

That said, `subscribe` is always present in the type of `controller.store`, so we could do this instead:

```suggestion
      if ('store' in controller && Boolean(controller.store?.subscribe)) {
```

This would also handle the extra edge cases where `store` or `subscribe` exist but are assigned undefined.

Applied here: https://github.com/MetaMask/metamask-extension/pull/28723/commits/26b87117353002423243a3abdb73fabf70592edc